### DISCUSSION TITLE
bam bam bhole! how, endpoint accessing on mobile browser i.e chrome LOCALLY using IP Address and PORT beacuse simple *Python* python - m http.server accessible using  having port 8000 accessible on mobile After Setting windows defender Inbound rule And allow app and feature defender setting but How FastAPI to access endpoint on mobile browser

### DISCUSSION THREAD

**Discussion by Karan09542** [fastapi/fastapi#11537] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
origins = [
    "http://127.0.0.1:5500",
    "http://localhost:8080",
    "http://localhost:5500",
    "http://127.0.0.1:5501",
# even i put this addres
    "http://computerIpAddr:5501",
    "mobileIpAddre:8000"
]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
# all requirement is imported AND all thing is work on computer but not on other machine like mobile phone browser

app = FastAPI()

@app.get("/bhola")
async def bholaBhola(prabhu: str= "bhola"):
    with open(f"{prabhu}Links.txt", "r") if  os.path.exists(f'{prabhu}Links.txt') else open("sambhuLinks.txt", "r") as bhola:
        bholebaba = bhola.read().split(" ")

    return secrets.choice(bholebaba)
```

### Description

how, endpoint accessing on mobile browser i.e chrome LOCALLY using IP Address and PORT beacuse simple *Python* python - m http.server accessible using  having port 8000 accessible on mobile After Setting windows defender Inbound rule And allow app and feature defender setting but How FastAPI to access endpoint on mobile browser
 
Acctually happing is that => i make a frontend => html and using javascript to fetch this enpoint and it work on computer but on mobile html and javascript work but FastAPI endpoint are not work

### Operating System

Windows

### FastAPI Version

0.104.1

### Pydantic Version

2.4.2

### Python Version

3.12.1

### Additional Context

no I used the GitHub search to find a similar question and didn't find it. * mean => i don't get any solution of that but i Get A
similar answer of question to another framework that is *Django* that have work on mobile(access endpoint on mobile)
site were we find it => link is - https://www.geeksforgeeks.org/how-to-access-pc-django-server-in-android/

**Comment by 0sahil** [comment]
Some quick things we can do to debug and run in dev environment:

1. Allowing all origins
```
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
```
2. Would like to get more info on how exactly are you running the app?
Try setting your host to be `0.0.0.0` and see if it works.

**Reply by Karan09542** [reply, marked as answer]
bam bam bhole! thankyou it work after setting host 0.0.0.0 and it run on mobile browser
but before it, it run on HOST 127.0.0.1 and PORT 8000 then why it not work

but a newThing is happenning that Endpoint is being accessing on mobile browser, mean fastapi api run on mobile BUT
when fetch in javascript only host 127.0.0.1 work rather then 0.0.0.0 and fixed 127 1st segment then changes 3 segment of dot number it also work i.e 127.10.0.1, 127.12.0.5, 127.0.10.5 so on

But running html, css , and js file on liveserver and access on mobile browser using computer IP Address and PORT 5501(live server host port) frontend work but FastAPI endpoint fetch not work using js

# Question Solution is solved:

Setting the host flag to 0.0.0.0
To access a FastAPI backend from a different machine/IP (than the local machine that is [running the server](https://stackoverflow.com/a/73909126/17865804)) on the same network, you would need to make sure that the host flag is set to [0.0.0.0](https://en.wikipedia.org/wiki/0.0.0.0). The IP address 0.0.0.0 means all IPv4 addresses on the local machine. If a host has two IP addresses, e.g., 192.168.10.2 and 10.1.2.5, and the server running on the host listens on 0.0.0.0, it will be reachable at both of those IPs. For example, through command line interface:

uvicorn main:app --host 0.0.0.0 --port 8000
or, programmatically:

if __name__ == '__main__':
    uvicorn.run(app, host='0.0.0.0', port=8000)
Note that RFC 1122 prohibits 0.0.0.0 as a destination address in IPv4 and only allows it as a source address, meaning that you can't type, for instance, http://0.0.0.0:8000 in the address bar of the browser and expect it to work. You should instead use one of the IPv4 addresses of the local machine, e.g., http://192.168.10.2:8000 (or, if you are testing the API on the same local machine running the server, you could use http://127.0.0.1:8000 or http://localhost:8000).

Adjusting Firewall Settings
You may also need to adjust your Firewall to allow external access to the port you specified, by creating an inbound firewall rule for Python. On Windows, this is usually created automatically, when allowing a program—Python, in this case—to communicate through Windows Firewall, and by default this allows traffic on Any port (for both TCP and UDP connections).

Adjusting CORS Settings
Additionally, if your frontend is listening on a separate IP address and/or port number from the backend, please make sure to have [CORS](https://fastapi.tiangolo.com/tutorial/cors/) enabled and properly configured, as desribed in [this answer](https://stackoverflow.com/a/71805329/17865804) and [this answer](https://stackoverflow.com/a/73963905/17865804). For example:

origins = ['http://localhost:3000','http://192.168.178.23:3000']

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


# Main Line

Making HTTP requests in JavaScript
Finally, please take a look at [this answer](https://stackoverflow.com/a/74407063/17865804) and [this answer](https://stackoverflow.com/a/74106637/17865804) regarding using the proper origin/URL when issuing a JavaScript fetch request from the frontend. In short, in your JavaScript asynchronous request, you should use the same domain name you typed in the address bar of your browser (but with the port number your backend server is listening on). If, for example, both the backend and the frontend server are listening on the same IP address and port number, e.g., 192.168.178.23:8000—that would be the case when using [Jinja2Templates](https://fastapi.tiangolo.com/advanced/templates/) for instance—you could access the frontend by typing in the address bar of the browser the url leading to the frontend page, e.g., http://192.168.178.23:8000/, and the fetch request should look like this:


further info : https://stackoverflow.com/questions/75040507/how-to-access-fastapi-backend-from-a-different-machine-ip-on-the-same-local-netw

# AND about 0.0.0.0 -
In FastAPI, the host 0.0.0.0 means that the application will listen on all network interfaces on the machine. This is in contrast to the default host, 127.0.0.1, which means that the application will only be accessible from the local machine.
command CLI: uvicorn main:app --host 0.0.0.0 --port 800

## to Specific IP address
command CLI : uvicorn main:app --host 192.168.1.100(like a example) --port 8000
